### PR TITLE
Add Git Sync Target

### PR DESCRIPTION
**Pull request by hammerdr**
This enables us to use a git repository and all of its
access to do a shared team repository for pet

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: sync/git.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+)
+
+type GitClient struct{}
+
+func NewGitClient() (Client, error) {
+	return GitClient{}, nil
+}
+
+func (g GitClient) GetSnippet() (*Snippet, error) {
+	directory, err := getDirectory()
```
Comment (knqyf263) [inline sync/git.go:20]: Handle error

File: sync/git.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	r, err := getRepository(directory)
```
Comment (knqyf263) [inline sync/git.go:21]: ditto

File: sync/git.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	w, err := r.Worktree()
```
Comment (knqyf263) [inline sync/git.go:22]: ditto

File: sync/git.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	})
+
+	return r.Push(&git.PushOptions{})
+}
+
+func getDirectory() (string, error) {
+	if config.Conf.Git.Directory != "" {
+		return config.Conf.Git.Directory, nil
+	}
+	return "", errors.New("Git Directory not found")
```
Comment (knqyf263) [inline sync/git.go:66]: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

```suggestion
	return "", errors.New("git directory not found")
```

File: sync/git.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	return r, nil
+}
+
+func getGitSignature() (string, string, error) {
+	if config.Conf.Git.Author != "" && config.Conf.Git.AuthorEmail != "" {
+		return config.Conf.Git.Author, config.Conf.Git.AuthorEmail, nil
+	}
+	return "", "", errors.New("Git Signature (Author, AuthorEmail) not found")
```
Comment (knqyf263) [inline sync/git.go:90]:
```suggestion
	return "", "", errors.New("git signature (Author, AuthorEmail) not found")
```

File: sync/git.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	r, err := git.PlainOpen(directory)
+
+	if err != nil {
+		r, err = git.PlainClone(directory, false, &git.CloneOptions{
+			URL:               config.Conf.Git.Repo,
+			RecurseSubmodules: git.DefaultSubmoduleRecursionDepth,
+		})
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "unable to create repository")
```
Comment (knqyf263) [inline sync/git.go:79]:
```suggestion
			return nil, fmt.Errorf("unable to create repository: %w", err)
```

Comment (knqyf263) [reply sync/git.go:79]: From Go 1.13 onwards, we can wrap an error with `fmt.Errorf`

File: sync/sync.go [hunk 7]
```diff
@@ -64,6 +64,13 @@ func NewSyncClient() (Client, error) {
 		}
 		return client, nil
 	}
+	if config.Conf.General.Backend == "git" {
+		client, err := NewGitClient()
+		if err != nil {
+			return nil, errors.Wrap(err, "Failed to initialize GitLab client")
```
Comment (knqyf263) [inline sync/sync.go:70]: ditto

**Review by knqyf263 — CHANGES_REQUESTED** [review]
Many errors are not handled. Could you fix it? Also, would you write a document in README? We should make it clear what kind of setting we expect.

**Comment by pwen** [comment]
+1 would be useful

**Comment by c33s** [comment]
can't wait for git sync support

**Pull request closed without merging**